### DISCUSSION TITLE
How to configure radioFSK for Lora FSK channel (EU)

### DISCUSSION THREAD

**Discussion by tauesdospi** [jgromes/RadioLib#794] [General]
Hi,

I have been trying to use the FSK mode of the lorawan stack for some FSK communication but it seems there are too my bugs and strange this so I'm giving up.

I'm new to RadioLib and trying to configure the radioFSK in order to communicate using the FSK channel specs defined in the Lorawan EU channel map.

I know these paramenters and configuring the lora fsk with this I'm able to send/receive packets from a Lorawan gateway using the FSK channel

#define FSK_FREQ    xxx // Hz
#define FSK_FDEV                                    25000     // Hz
#define FSK_DATARATE                                50000     // bps
#define FSK_BANDWIDTH                               50000     // Hz
#define FSK_AFC_BANDWIDTH                           83333     // Hz
#define FSK_PREAMBLE_LENGTH                         5         // Same for Tx and Rx
#define FSK_FIX_LENGTH_PAYLOAD_ON                   false

I have tried this in RadioLib

  SX1276 radio = new Module(GPIO_NUM_18, GPIO_NUM_26, GPIO_NUM_14, GPIO_NUM_35);
radio.beginFSK(868.8, 50, 25, 83.3, 17, , false);

I guess I'm missing several things (syncwords? others? ) as Radiolib is a generic all-purpourse lib and the lorawan fsk functions are directed to work only with their specs.

I will appreciate any help on how to setup RadioLib to be able to comunicate to Lorawan EU868 FSK channels.

Thank you
Antonio

**Comment by jgromes** [comment]
At the moment, LoRaWAN support in RadioLib is still work-in-progress. The node starts in LoRa mode, and because adaptive data rate was not yet implemented, the node will never enter the FSK channels. So for now, there's no support for LoRaWAN FSK channels, though it is being worked on.

However, as far as I understand LoRaWAN, you don't really select that you want to use the FSK channel, instead it will be the highest possible transfer speed when using the adaptive data rate feature. Though it is possible I am mistaken, and if you perform OTA activation on the FSK channel with ADR disabled, the node will remain there. Guess I can test this

**Reply by tauesdospi** [reply]
Hi ! Thank you for the quick answer.

Is it possible I didn't explained correctly but I don't want anything about Lora modulation o Lorawan protocol (no join no ota etc), just configure radiolib to comunicate raw fsk packets to a lorawan gateway which has a fsk channel. 

The question is I don't know  exactly how to configure all radiolib fsk settings in order to comunicate with the fsk modem all lorawan gateways  have. I know the frequency, bandwidth, baud rate, preamble length, variable payload size, etc.. but it seems others parameters have to be set to communicate.

I'm able to send/receive raw fsk packets from the lorawan gateway using the fsk channel (I'm using a propietary p2p protocol) and I want to use radiolib in the equipment side to send/receive raw fsk packets.

Thank you,
Antonio

**Reply by jgromes** [reply]
> I'm able to send/receive raw fsk packets from the lorawan gateway using the fsk channel (I'm using a propietary p2p protocol)

How are you doing that? Is your gateway using some custom software? If it's a LoRaWAN gateway I would expect it to be using the LoRaWAN stack. Perhaps better question is what exactly do you mean by LoRaWAN gateway?

> The question is I don't know exactly how to configure all radiolib fsk settings in order to comunicate with the fsk modem

There's the [SX127x FSK Modem example](https://github.com/jgromes/RadioLib/blob/master/examples/SX127x/SX127x_FSK_Modem/SX127x_FSK_Modem.ino), which shows all the basic configuration. In addition, there's full API reference for the SX127x module here: https://jgromes.github.io/RadioLib/class_s_x127x.html

**Reply by tauesdospi** [reply]
"How are you doing that? Is your gateway using some custom software?"

The lorawan gateway is a dumb packet sender/receiver with lora o fsk modulations.  I can communicate with it with the "lora pkt forwarder udp protocol" .. basically I sent to the IP:port an UDP packet in JSON format with the frequency, modulation, bandwidth, etc. and the payload and it sends to the air. Also when the gateway receives any packet in any of the configured channels with any of the configured modulations, it sends back to my IP:port also in JSON format. So it is a very easy way to have a profesional (enclosure, antenna, etc) generic radio modem which can talk lora o fsk.

I have seen the FSK examples and tries several parameters but I'm unable to commnicate. I don't know which settings are configured for the FSK lora channel like  setCurrentLimit? setDataShaping ?setSyncWord? setOOK ? My initial question was regarding the exact configuration of all these parameters in order to communicate as I don't know what kind of settings are configured in the FSK lora gateway channel beyond frequency, bandwidth, baudrate, preamble length...

Thanks,
Antonio

**Reply by jgromes** [reply]
> The lorawan gateway is a dumb packet sender/receiver with lora o fsk modulations

I think I understand now - when you write "lorawan gateway" you mean that you have the SX1302 concentrator with some frontends runings this: https://github.com/Lora-net/packet_forwarder but it's not connected to e.g. TTN.

> I don't know which settings are configured for the FSK lora channel

Well neither do I, sorry. That seems like a question you should ask either the gateway supplier. RadioLib lets you configure any of the settings available to the SX127x, but the actual values are defined by the gateway, I assume.

**Reply by tauesdospi** [reply]
Exactly... lots of commercial lorawan gateways are sx130x concentrators with the packet forwarder software.. like mikrotik wap lora, multitech conduit, etc.. some more expensive also have the lorawan stack and network server inside (eg. milesight) but all of the can be communicated directly by the packet-forwarder protocol.

Thanks for your time! I will continue investigating in order to figure what set of parameters are required to communicate to the FSK channel of such radios... It should be specified at some place as it is standarized across all lorawan gateways but I can't find it.. Will let you know !

Best regards
Antonio

**Reply by tauesdospi** [reply]
got it ! I'm posting it here in case anybody else requires this at some point:

This setup permits raw FSK tx/rx with any lorawan gateway (packet-forwarder) in EU region. For other regions frequency and some other parameter may change.

radio.beginFSK();
radio.setFrequency(868.8);  // this is the standarized EU868 Lorawan FSK channel
radio.setBitRate(50.0);
radio.setFrequencyDeviation(25);
radio.setRxBandwidth(50.0);
radio.setOutputPower(17.0); // or your power
radio.setCurrentLimit(100); // may not be required
radio.setDataShaping(RADIOLIB_SHAPING_1_0);
 uint8_t syncWord[] = {0xC1, 0x94, 0xC1 };
radio.setSyncWord(syncWord, 3);
radio.disableAddressFiltering();
radio.setEncoding(RADIOLIB_ENCODING_WHITENING); 
radio.setCRC(true);
radio.variablePacketLengthMode();

**Reply by jgromes** [reply]
Thanks for poisting the parameters - this gives me an idea that I could extend the LoRaWAN implementation to allow FSK usage for non-LoRa modules. So one could connect plain old FSK transceiver to a LoRaWAN network, that would be rather interesting.

**Reply by jgromes** [reply]
@tauesdospi where did you find the frequency for the EU868 FSK channel? It's not listed in the 1.1 regional parameters document.

**Reply by tauesdospi** [reply]
It is noted in several places eg. https://www.thethingsnetwork.org/docs/lorawan/frequency-plans/

the exact parameters to use it are:

 state = radio.beginFSK(); 
  state = radio.setFrequency(868.8);
  state = radio.setBitRate(50.0);
  state = radio.setFrequencyDeviation(25);
  state = radio.setRxBandwidth(50.0);
  state = radio.setAFCBandwidth(83.3);
  state = radio.setDataShaping(RADIOLIB_SHAPING_1_0);
  uint8_t syncWord[] = {0xC1, 0x94, 0xC1 };
  state = radio.setSyncWord(syncWord, 3);
  state = radio.setPreambleLength(40);
  state = radio.disableAddressFiltering();
  state = radio.setEncoding(RADIOLIB_ENCODING_WHITENING);
  state = radio.setCRC(true);
  state = radio.variablePacketLengthMode();

regards
Antonio
